r/src/compute.cpp

```diff
@@ -208,7 +208,7 @@ std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(
     return out;
   }
 
-  if (func_name == "hash_count_distinct") {
+  if (func_name == "count_distinct" || func_name == "hash_count_distinct") {
```

Comment:
Unrelated to this PR, but it's amusing that R tests for function names even though the name of the options class is available as `FunctionDoc::options_class`. @nealrichardson @jonkeane @thisisnic

Reply:
Sorry, can you rephrase the question/suggestion @pitrou , I'm not sure what you're asking?

Reply:
It's more of a suggestion than a question. Instead of manually mapping function names to options classes in the R code, just use the facility exported by the C++ APIs:
https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/function.h#L139-L140

You can get the function object directly from the registry:
https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/registry.h

Reply:
The rough idea:
```c++

std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(
    const std::string& options_name, cpp11::list options) {
  if (options_name == "ScalarAggregateOptions") {
    using Options = arrow::compute::ScalarAggregateOptions;
    auto out = std::make_shared<Options>(Options::Defaults());
    if (!Rf_isNull(options["min_count"])) {
      out->min_count = cpp11::as_cpp<int>(options["min_count"]);
    }
    if (!Rf_isNull(options["skip_nulls"])) {
      out->skip_nulls = cpp11::as_cpp<bool>(options["skip_nulls"]);
    }
    return out;
  }
  if (options_name == "TDigestOptions") {
    // etc;
  }
  return nullptr;
}

// [[arrow::export]]
SEXP compute__CallFunction(const std::string& func_name, cpp11::list args, cpp11::list options) {
  auto func = ValueOrStop(
      arrow::compute::GetFunctionRegistry()->GetFunction(func_name));
  auto opts = make_compute_options(func.doc()->options_class, options);
  auto datum_args = arrow::r::from_r_list<arrow::Datum>(args);
  auto out = ValueOrStop(func->Execute(datum_args, opts.get(), gc_context()));
  return from_datum(std::move(out));
}
```

Reply:
Definitely!